jcl/src/java.base/share/classes/java/lang/invoke/VarHandle.java

```diff
@@ -1196,8 +1205,41 @@ public String toString() {
 		 * @return ClassDesc of variable type
 		 */
 		public ClassDesc varType() {
-			return constantType();
+			return (type == Kind.ARRAY) ? getArrayTypeClassDesc().componentType() : getFieldTypeClassDesc();
+		}
+
+		/* helper method. declaring class is always the first bootstrap argument for a field type
+		 * @throws InternalError if VarHandleDesc was created improperly
+		 */
+		private ClassDesc getFieldDeclaringClassDesc() {
+			ConstantDesc[] args = bootstrapArgs();
+			if ((args.length < 2) || !(args[0] instanceof ClassDesc)) {
+				throw new InternalError();
+			}
+			return (ClassDesc)args[0];
+		}
+
+		/* helper method. fieldType is always the second bootstrap argument for a field type
+		 * @throws InternalError if VarHandleDesc was created improperly
+		 */
+		private ClassDesc getFieldTypeClassDesc() {
+			ConstantDesc[] args = bootstrapArgs();
+			if ((args.length < 2) || !(args[1] instanceof ClassDesc)) {
+				throw new InternalError();
+			}
+			return (ClassDesc)args[1];
 		}
+
+		/* helper method. array component type is always the first bootstrap argument for an array type
+		 * @throws InternalError if VarHandleDesc was created improperly
+		 */
+		private ClassDesc getArrayTypeClassDesc() {
+			ConstantDesc[] args = bootstrapArgs();
+			if ((args.length == 0) || !(args[0] instanceof ClassDesc) || (type != Kind.ARRAY)) {
+				throw new InternalError();
+			}
+			return (ClassDesc)args[0];
+		}		
```

Comment:
All three of these methods need their javadoc cleaned up and messages added to their exceptions